spec/syntax.md

```diff
@@ -148,19 +148,39 @@ MAY include an _annotation_ that is applied to the external value.
 
 A **_<dfn>local-declaration</dfn>_** binds a _variable_ to the resolved value of an _expression_.
 
-Declared _variables_ MUST NOT be used before their _declaration_,
-and their values MUST NOT be self-referential;
-otherwise, a _message_ is not considered _valid_.
-
-Multiple _declarations_ MUST NOT bind a value to the same _variable_;
-otherwise, a _message_ is not considered _valid_.
-
 ```abnf
 declaration = input-declaration / local-declaration
 input-declaration = input [s] variable-expression
 local-declaration = local s variable [s] "=" [s] expression
 ```
 
+_Variables_, once declared, MUST NOT be redeclared. 
+A _message_ that does any of the following is not _valid_ and will produce a 
+Duplicate Declaration error during formatting:
+- An _input-declaration_ MUST NOT bind a _variable_ that appears as a _variable_ in a previous 
+  _declaration_.
+- A _local-declaration_ MUST NOT bind a _variable_ that appears as a _variable_ in a previous
+  _declaration_.
+- A _local-declaration_ MUST NOT bind a _variable_ that appears in its _expression_.
+
+A _local-declaration_ MAY overwrite an external input value as long as the
+external input value does not appear in a _declaration_.
+
+> [!Note]
+> These restrictions only apply to _declarations_.
+> A _placeholder_ or _selector_ MAY supply a different annotation than one
+> provided in a _declaration_ for the same _variable_.
+> For example, this message is _valid_:
+> ```
+> {{
+>    input {$var :number maxFractionDigits=0}
+>    match {$var :plural maxFractionDigits=2}
+>    when 0 {{The selector can re-annotate {$var}}}
+>    when * {{This pattern can re-annotate {$var :number maxFractionDigits=3}}}
```

Comment:
The exposition is better, but I'm still confused about the semantics. In the scope of the `input` declaration (the scope where `$var` is defined), does `$var` refer to the result of applying the `:number` formatter to `$var` with the option `maxFractionDigits` bound to 0? Or does it just refer to whatever value was passed in for `$var`? If the latter is true, what's the point of applying the `:number` annotation in the `input` declaration? And if the former is true, what does it mean to apply the `:number` formatter to an already-formatted number (or is `:number` expected to check whether its argument is a raw number or a formatted number, and handle those cases differently?)

Reply:
Our specification is explicit in that it doesn't say whether the declaration is early or late binding, which I think means that it doesn't matter. Another way to say that is that the `input` declaration means that any reference to `$var` should be interpreted through the function `:number` and the passed options.

Maybe the word `re-annotate` is not right? Perhaps:

```suggestion
>    input {$var :number maxFractionDigits=0}
>    match {$var :plural maxFractionDigits=2}
>    when 0 {{The selector can apply a different annotation to {$var} for the purposes of selection}}
>    when * {{A placeholder in a pattern can apply a different annotation to {$var :number maxFractionDigits=3}}}
```

Reply:
> And if the former is true, what does it mean to apply the :number formatter to an already-formatted number (or is :number expected to check whether its argument is a raw number or a formatted number, and handle those cases differently?)

We don't specify what type exactly the argument to `:number` should be, but we do say that (a) it should be the *resolved* value, and that (b) the shape of the resolved value is implementation-specific. If a "resolved value" is allowed by the implementation to be a raw value, then functions must accept such raw values as arguments, in addition to richer "formattable" data structures. 

Also, note that the same situation can be reproduced without `input`, using multiple `local` declarations:

```
local $a = {1 :number minFractionDigits=2} // formats as 1.00
local $b = {$a :number minFractionDigits=3} // formats as 1.000
```

Reply:
Aside: I'm not sure if the spec is currently clear enough about the requirement to allow functions to inspect the raw value and the formatting options of their arguments. Specifically, while a function may eagerly format its operand to a string, it should (must?) also return the raw underlying value and the formatting options used for formatting, in case another function wants to extend them or use them for other logic. Consider:

```
local $a = {1 :number minIntegerDigits=3} // formats as 001.
local $b = {$a :number minFractionDigits=3} // formats as 001.000
// min integer digits are preserved from the previous call.
```

```
input {$item :noun case=accusative count=1}
local $colorMatchingGrammaticalNumberGenderCase = {$color :adjective accord=$item}
// :adjective inspects $item to learn about case=accusative, count=1
```

Should I file an issue about this?